packages/datadog-instrumentations/src/apollo-gateway.js

```diff
@@ -0,0 +1,110 @@
+const {
+  addHook,
+  channel
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+const tracingChannel = require('dc-polyfill').tracingChannel
+
+const validateCh = tracingChannel('apm:apollo-gateway:validate')
+const requestCh = tracingChannel('apm:apollo-gateway:request')
+const fetchCh = tracingChannel('apm:apollo-gateway:fetch')
+const planCh = tracingChannel('apm:apollo-gateway:plan')
+const executeCh = tracingChannel('apm:apollo-gateway:execute')
+const postProcessingCh = tracingChannel('apm:apollo-gateway:postprocessing')
+
+const generalErrorCh = channel('apm:apollo-gateway:general:error')
+const REQUEST_CTX = {}
+
+function wrapExecutor (executor) {
+  return function (...args) {
+    REQUEST_CTX.requestContext = args[0]
+    REQUEST_CTX.gateway = { ...this }
+    return executor.apply(this, args)
+  }
+}
+
+function wrapApolloGateway (ApolloGateway) {
+  class ApolloGatewayWrapper extends ApolloGateway {
+    constructor (...args) {
+      super(...args)
+      shimmer.wrap(this, 'executor', wrapExecutor)
+    }
+  }
+  return ApolloGatewayWrapper
+}
+
+function wrapRecordExceptions (recordExceptions) {
+  return function wrappedRecordExceptions (...args) {
+    const errors = args[1]
+    // only the last exception in the array of exceptions will be reported on the span,
+    // this is mimicking apollo-gateways internal instrumentation
+    // TODO: should we consider a mechanism to report all exceptions? since this method aggregates all exceptions
+    // where as a span can only have one exception set on it at a time
+    generalErrorCh.publish({ error: errors[errors.length - 1] })
+    return recordExceptions.apply(this, args)
+  }
+}
+
+function wrapStartActiveSpan (startActiveSpan) {
+  return function (...args) {
+    const firstArg = args[0]
+    if (typeof firstArg !== 'string') return startActiveSpan.apply(this, args)
+
+    const cb = args[args.length - 1]
+    let newCb
+    switch (firstArg) {
+      case 'gateway.request': {
+        newCb = function (...callbackArgs) {
+          return requestCh.tracePromise(cb, REQUEST_CTX, this, ...callbackArgs)
+        }
+        break
+      }
+      case 'gateway.plan' : {
+        newCb = function (...callbackArgs) {
+          return planCh.traceSync(cb, {}, this, ...callbackArgs)
+        }
+        break
+      }
+      case 'gateway.validate': {
+        newCb = function (...callbackArgs) {
+          return validateCh.traceSync(cb, {}, this, ...callbackArgs)
+        }
+        break
+      }
+      case 'gateway.execute': {
+        newCb = function (...callbackArgs) {
+          return executeCh.tracePromise(cb, {}, this, ...callbackArgs)
+        }
+        break
+      }
+      case 'gateway.fetch': {
+        newCb = function (...callbackArgs) {
+          return fetchCh.tracePromise(cb, { attributes: args[1].attributes }, this, ...callbackArgs)
```

Comment:
Shouldn't this be capturing from `callbackArgs.attributes` instead? And if so, shouldn't there be a test for this?

Reply:
tracer.startActivateSpan has attributes available to it, not the callback which only has span available to it in args.